helpers/google_helpers.go

```diff
@@ -0,0 +1,195 @@
+	}
+	if filter.EventSource != "" {
+		f += " AND " + eventSource + " LIKE '" + filter.EventSource + "'"
+	}
+	if filter.EventType != "" {
+		f += " AND " + eventType + " LIKE '" + filter.EventType + "'"
+	}
+	return f
+}
+
+// getTableName sets the bigquery table name.
+// Each environment has different table name. The default points to the develop audit log table.
+// Currently only develop and staging is allowed.
+func getTableName() string {
+	projectID := getProjectID()
+	env := "dev"
+	runEnv := os.Getenv("RUN_ENV")
+	if runEnv == "staging" {
+		env = "stg"
+	}
+	return projectID + "." + env + "_audit_log." + env + "_audit_log"
+}
+
+func processBqResults(iter *bigquery.RowIterator) (ReturnValues, error) {
+	var row ReturnValues
+	for {
+		err := iter.Next(&row)
+		if e.Is(err, iterator.Done) {
+			return row, nil
+		}
+		if err != nil {
+			return row, err
+		}
+	}
+}
+
+// FillFilterFieldsFromEnvironment creates a FilterFields object and prefills the App space information from the
+// App credential environment variable.
+func FillFilterFieldsFromEnvironment() (FilterFields, error) {
+	applicationCred := os.Getenv("INDYKITE_APPLICATION_CREDENTIALS")
+	var jsonMap map[string]any
+	var f FilterFields
+	if applicationCred == "" {
+		return f, errors.NewInvalidArgumentError("Missing INDYKITE_APPLICATION_CREDENTIALS environment variable")
+	}
+	err := json.Unmarshal([]byte(applicationCred), &jsonMap)
+	if err != nil {
+		return f, err
+	}
+	appSpaceID := fmt.Sprintf("%v", jsonMap["appSpaceId"])
+	f = FilterFields{
+		AppSpaceID: appSpaceID,
+		RowNumber:  "1",
+	}
+
+	return f, nil
+}
+
+// BqClient creates a client connection to BigQuery.
+func BqClient(ctx context.Context) (*bigquery.Client, error) {
+	client, err := bigquery.NewClient(ctx, getProjectID())
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+// QueryAuditLog sends a query request to the BigQuery with the given parameters and returns with its result
+// It returns with error if the RUN_ENV environment variable is not set or not develop/staging.
+func QueryAuditLog(ctx context.Context, client *bigquery.Client, fFields *FilterFields) (ReturnValues, error) {
+	var params bqQueryParameters
+	params.tableName = getTableName()
+	if fFields.RowNumber != "" {
+		params.RowNumber = fFields.RowNumber
+	}
+	params.filters = getFilterFields(fFields)
+	params.ReturnFields = getReturnFields()
+	for {
+		select {
+		case <-ctx.Done():
+			return ReturnValues{}, e.New("the context is done")
+		case <-time.After(time.Second):
+			query := client.Query(
+				"SELECT " +
+					params.ReturnFields +
+					" FROM `" +
+					params.tableName +
+					"` " +
+					params.filters +
+					" ORDER BY publish_time DESC LIMIT " +
+					params.RowNumber + ";")
+			iter, err := query.Read(ctx)
+			if err != nil {
+				return ReturnValues{}, err
+			}
+			res, err := processBqResults(iter)
+			if err != nil {
+				return ReturnValues{}, err
+			}
+			return res, nil
+		}
+	}
```

Comment:
Why the outer infinite loop? It will never run more than once

Reply:
Yeh, it was in a different place. Now it is not necessary. Removed.